common/src/main/java/com/wynntils/handlers/inventory/InventoryHandler.java

```diff
@@ -0,0 +1,449 @@
+        ItemStack newStack = event.getItemStack();
+        pending.removeIf(
+                entry -> now >= entry.expiryTime || !entry.interaction.onSlotUpdate(containerId, slotNum, newStack));
+    }
+
+    @SubscribeEvent
+    public void onWorldStateChange(WorldStateEvent event) {
+        running = null;
+        pending.clear();
+    }
+
+    public void forceSync(AbstractContainerMenu menu) {
+        McUtils.sendPacket(new ServerboundContainerClickPacket(
+                menu.containerId,
+                menu.getStateId(),
+                Math.min(1, menu.slots.size() - 1),
+                0,
+                ClickType.SWAP,
+                new ItemStack(Items.BARRIER, 31),
+                Int2ObjectMaps.emptyMap()));
+    }
+
+    private void forceSyncLater(AbstractContainerMenu menu) {
+        Managers.TickScheduler.scheduleNextTick(() -> forceSync(menu));
+    }
+
+    private void expect(PendingInteraction interaction) {
+        pending.add(new PendingInteractionEntry(interaction));
+    }
+
+    private static boolean areItemsSimilar(ItemStack a, ItemStack b) {
+        if (a.isEmpty()) {
+            return b.isEmpty();
+        } else {
+            return !b.isEmpty()
+                    && a.getHoverName().getString().equals(b.getHoverName().getString());
+        }
+    }
+
+    private static boolean areStacksSimilar(ItemStack a, ItemStack b) {
+        return areItemsSimilar(a, b) && a.getCount() == b.getCount();
+    }
+
+    private final class RunningInteraction {
+        private final AbstractContainerMenu menu;
+        private final boolean single;
+        private final ItemStack heldStack;
+        private final IntList slots = new IntArrayList();
+
+        private RunningInteraction(AbstractContainerMenu menu, boolean single, ItemStack heldStack) {
+            this.menu = menu;
+            this.single = single;
+            this.heldStack = heldStack;
+        }
+
+        private boolean handle(ContainerClickEvent event) {
+            if (event.getClickType() != ClickType.QUICK_CRAFT
+                    || event.getContainerMenu().containerId != menu.containerId) {
+                running = null;
+                return false;
+            }
+
+            int mask = event.getMouseButton();
+            if (((mask & 0x4) == 0) == single) {
+                running = null;
+                return false;
+            }
+
+            switch (event.getMouseButton() & 0x3) {
+                case 1 -> slots.add(event.getSlotNum());
+                case 2 -> {
+                    forceSyncLater(menu);
+                    expect(new PendingInteraction.Spread(menu, slots, heldStack, single));
+                    running = null;
+                }
+                default -> {
+                    running = null;
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    private interface PendingInteraction {
```

Comment:
This file is rather long, and the majority of it is nested classes. I don't advocate moving every single nested class into a top-level class, but perhaps try to move the PendingInteraction hierarchy to a separate file, and see if it makes the files more balanced?